src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    internal static class SerializationThreadPool
+    {
+        public static Task<object?> RunOnBackgroundThreadAsync(Func<object?> start)
+            => ImmediateBackgroundThreadPool.QueueAsync(start);
+
+        public static Task<object?> RunOnBackgroundThreadAsync(Func<object?, object?> start, object? obj)
+            => ImmediateBackgroundThreadPool.QueueAsync(start, obj);
+
+        /// <summary>
+        /// Naive thread pool focused on reducing the latency to execution of chunky work items as much as possible.
+        /// If a thread is ready to process a work item the moment a work item is queued, it's used, otherwise
+        /// a new thread is created. This is meant as a stop-gap measure for workloads that would otherwise be
+        /// creating a new thread for every work item.
+        /// </summary>
+        /// <remarks>
+        /// This class is derived from <see href="https://github.com/dotnet/machinelearning/blob/ebc431f531436c45097c88757dfd14fe0c1381b3/src/Microsoft.ML.Core/Utilities/ThreadUtils.cs">dotnet/machinelearning</see>.
+        /// </remarks>
+        private static class ImmediateBackgroundThreadPool
+        {
+            /// <summary>How long should threads wait around for additional work items before retiring themselves.</summary>
+            private static readonly TimeSpan s_idleTimeout = TimeSpan.FromSeconds(1);
+
+            /// <summary>The queue of work items. Also used as a lock to protect all relevant state.</summary>
+            private static readonly Queue<(Delegate, object?, TaskCompletionSource<object?>)> s_queue = new();
+
+            /// <summary>The number of threads currently waiting in <c>tryDequeue</c> for work to arrive.</summary>
+            private static int s_availableThreads = 0;
+
+            /// <summary>
+            /// Queues an <see cref="Action"/> delegate to be executed immediately on another thread,
```

Comment:
I didn't see use of `Action` of the implementation. Should these references actually be to `Func<T>`/`Func<T1, T2>`?

Reply:
➡️ Now updated